### PR TITLE
Adding Proxy Endpoint Unifier

### PR DESCRIPTION
**Pull request by anaik91**
## Description

Apigee X has a limitation of hosting only 5 Proxy Endpoints per proxy.
Apigee OPDK /Edge has no such limitation.
Objective is take a proxy bundle and smartly convert them into conditional flows and group them with other proxy endpoints.
## Housekeeping
(please check all that apply [x], do not edit the text)
- [x] I have run all the tests locally and they all pass.
- [x] I have followed the relevant style guide for my changes.

## Full Repo Validation Required
(please check all that apply [x], do not edit the text)
- [ ] PR requires full pipeline run (Run for changes only by default).

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

**Comment by anaik91** [comment]
@danistrebel @OmidTahouri : kindly  run gcbrun when you get a chance

**Review by OmidTahouri — CHANGES_REQUESTED** [review]
Partial review

File: tools/proxy-endpoint-unifier/utils.py [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+def parse_proxy_root(dir):
+    file = get_proxy_entrypoint(dir)
+    if file is None:
+        return {}
+    doc = parse_xml(file)
+    return doc
+
+
+def read_proxy_artifacts(dir, entrypoint):
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/utils.py:126]: What about resources?

**Review by OmidTahouri — CHANGES_REQUESTED** [review]

File: tools/proxy-endpoint-unifier/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+## Pre-Requisites
+* python3.x
+* Please Install required Python Libs 
+```
+    python3 -m pip install requirements.txt
+```
+* Please fill in `input.properties`
+```
+    [common]
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/README.md:18]: I believe these are sorted in https://github.com/apigee/devrel/pull/687

File: tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/manifests/manifest.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/manifests/manifest.xml:1]: Acknowledged !

File: tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/policies/Populate-Cache-1.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    limitations under the License.
+-->
+<PopulateCache async="false" continueOnError="false" enabled="true" name="Populate-Cache-1">
+    <DisplayName>Populate Cache-1</DisplayName>
+    <Properties/>
+    <CacheKey>
+        <Prefix>my </Prefix>
+        <KeyFragment>test</KeyFragment>
+    </CacheKey>
+    <CacheResource>test cache</CacheResource>
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/policies/Populate-Cache-1.xml:21]: Acknowledged

File: tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/targets/default.xml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        <Response>
+            <Step>
+                <Name>ExtractVariables-3</Name>
+            </Step>
+        </Response>
+    </PostFlow>
+    <Flows/>
+    <HTTPTargetConnection>
+        <Properties/>
+        <URL>http://mocktarget.apigee.net/</URL>
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/test/api_bundles/apiproxy/targets/default.xml:32]: Acknowledged

File: tools/proxy-endpoint-unifier/utils.py (hunk 1 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/utils.py:126]: For this MVP we are only dealing with Policies and proxy endpoints

File: tools/proxy-endpoint-unifier/main.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                'ProxyEndpoints': []
+            }
+            for each_path, pes in each_group.items():
+                each_pe = '-'.join(pes)
+                merged_pes[each_pe] = utils.merge_proxy_endpoints(
+                    processing_final_dict[each_api],
+                    each_path,
+                    pes
+                )
+                merged_objects[f"{each_api}_{index}"]['Name'] = f"{final_dict[each_api]['proxyName']}_{index}"  # noqa
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/main.py:81]: For this MVP we are only dealing with Policies and proxy endpoints

File: tools/proxy-endpoint-unifier/pipeline.sh [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+# Install Dependencies
+python3 -m pip install -r "$SCRIPTPATH/requirements.txt"
+
+# Generate Gcloud Acccess Token
+APIGEE_ACCESS_TOKEN=$(gcloud auth print-access-token)
+export APIGEE_ACCESS_TOKEN
+
+# Execute Utility
+python3 "$SCRIPTPATH/main.py"
```
[the remark this answers is not in the document]
Comment (anaik91) [reply tools/proxy-endpoint-unifier/pipeline.sh:48]: If the script fails with non zero exit code , I am expecting the `pipeline.sh` to fail

File: tools/proxy-endpoint-unifier/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (OmidTahouri) [reply tools/proxy-endpoint-unifier/main.py:82]: Let's add a comment to the README to state that this is a currently known limitation

File: tools/proxy-endpoint-unifier/main.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        'merged_pes': merged_pes,
+        'merged_objects': merged_objects,
+    }
+    if export_debug_file:
+        utils.export_debug_log(files)
+
+    if validation_enabled:
+        errors = {}
+        gcp_project_id = cfg['validate']['gcp_project_id']
+        x = ApigeeXorHybrid(gcp_project_id)
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/main.py:116]: Rename class to Apigee

File: tools/proxy-endpoint-unifier/pipeline.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+rm -rf "$SCRIPTPATH/input.properties"
+rm -rf "$SCRIPTPATH/transformed"
+rm -rf "$SCRIPTPATH/transformed_bundles"
+
+# Generate input file
+cat > "$SCRIPTPATH/input.properties" << EOF
+[common]
+input_apis=$SCRIPTPATH/test/api_bundles
+processed_apis=$SCRIPTPATH/transformed
+proxy_bundle_directory=$SCRIPTPATH/transformed_bundles
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/pipeline.sh:31]: zipped_bundles

File: tools/proxy-endpoint-unifier/pipeline.sh [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+proxy_endpoint_count=4
+debug=true
+
+[validate]
+enabled=true
+gcp_project_id=$APIGEE_ORG
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/pipeline.sh:37]: Use `APIGEE_X_ORG`, per https://github.com/apigee/devrel/blob/main/PIPELINES.md

File: tools/proxy-endpoint-unifier/utils.py [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+            merged_pe['ProxyEndpoint']['PreFlow']['Request']['Step'].extend(
+                process_steps(each_pe_info['ProxyEndpoint']['PreFlow']['Request'],condition)  # noqa
+            )
+            merged_pe['ProxyEndpoint']['PreFlow']['Response']['Step'].extend(
+                process_steps(each_pe_info['ProxyEndpoint']['PreFlow']['Request'],condition)  # noqa
+            )
+            merged_pe['ProxyEndpoint']['PostFlow']['Request']['Step'].extend(
+                process_steps(each_pe_info['ProxyEndpoint']['PostFlow']['Request'],condition)  # noqa
+            )
+            merged_pe['ProxyEndpoint']['PostFlow']['Response']['Step'].extend(
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/utils.py:450]: What about the `PostClientFlow`?

File: tools/proxy-endpoint-unifier/xorhybrid.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import requests
+import os
+
+
+class ApigeeXorHybrid:
```
Comment (OmidTahouri) [inline tools/proxy-endpoint-unifier/xorhybrid.py:21]: Rename class and file to just `Apigee`, as discussed

**Review by OmidTahouri — CHANGES_REQUESTED** [review]

File: tools/proxy-endpoint-unifier/pipeline.sh (hunk 9 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/pipeline.sh:31]: done

File: tools/proxy-endpoint-unifier/pipeline.sh (hunk 10 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/pipeline.sh]: done

File: tools/proxy-endpoint-unifier/utils.py (hunk 11 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/utils.py:459]: Added support

File: tools/proxy-endpoint-unifier/xorhybrid.py (hunk 12 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/xorhybrid.py]: renamed

File: tools/proxy-endpoint-unifier/main.py (hunk 8 above)
Comment (anaik91) [reply tools/proxy-endpoint-unifier/main.py]: renamed

**Review by OmidTahouri — APPROVED** [review]
LGTM

File: tools/proxy-endpoint-unifier/README.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+## Usage
+Run the script as below
+```
+python3 main.py
+```
+
+## Limitations
+* This tool does not currently handle the resources within API proxies.
+
+## Copyright
```
Comment (danistrebel) [inline tools/proxy-endpoint-unifier/README.md:46]: can we remove this as the solution inherits the license from the repo level?

**Review by danistrebel — APPROVED** [review]
LGTM, if we can remove the copyright note on the README, that would be perfect.

File: tools/proxy-endpoint-unifier/README.md (hunk 13 above)
Comment (OmidTahouri) [reply tools/proxy-endpoint-unifier/README.md]: +1

**Review by OmidTahouri — APPROVED** [review]

**Pull request merged**